common/predictive-text/unit_tests/headless/edit-distance/distance-modeler.js

```diff
@@ -222,4 +222,149 @@ describe.only('Correction Distance Modeler', function() {
+
+    before(function() {
+      testModel = new models.TrieModel(jsonFixture('tries/english-1000'));
+    });
+
+    let checkResults_teh = function(iter) {
+      let firstSet = iter.next();  // {value: <actual value>, done: <iteration complete?>}
+      assert.isFalse(firstSet.done);
+      
+      firstSet = firstSet.value; // Retrieves <actual value>
+      assert.equal(firstSet[1], 1);
+      assert.equal(firstSet[0].length, 1); // A single sequence ("ten") should be the best match.
+
+      let chars = firstSet[0][0].map(value => value.key);
+      assert.equal(chars.join(''), "ten");
+
+      let secondBatch = [
+        'beh',  'te',  'tec',
+        'tech', 'tel', 'tem',
+        'ter',  'tes', 'th',
+        'the'
+      ];
+
+      let secondSet = iter.next();  // {value: <actual value>, done: <iteration complete?>}
+      assert.isFalse(secondSet.done);
+      
+      secondSet = secondSet.value; // Retrieves <actual value>
+      assert.equal(secondSet[1], 1.5);
+      assert.equal(secondSet[0].length, secondBatch.length);
+
+      let entries = secondSet[0].map(function(sequence) {
+        return sequence.map(value => value.key).join('');
+      }).sort();
+      assert.deepEqual(entries, secondBatch);
+
+      let thirdBatch = [
+        'cen', 'en',  'gen',
+        'ken', 'len', 'men',
+        'sen', 'tha', 'then',
+        'thi', 'tho', 'thr',
+        'thu', 'wen'
+      ];
+
+      let thirdSet = iter.next();  // {value: <actual value>, done: <iteration complete?>}
+      assert.isFalse(thirdSet.done);
+      
+      thirdSet = thirdSet.value; // Retrieves <actual value>
+      assert.equal(thirdSet[1], 2);
+      assert.equal(thirdSet[0].length, thirdBatch.length);
+
+      entries = thirdSet[0].map(function(sequence) {
+        return sequence.map(value => value.key).join('');
+      }).sort();
+      assert.deepEqual(entries, thirdBatch);
+    }
+
+    it('Simple search (paralleling "Small integration test")', function() {
+      // The combinatorial effect here is a bit much to fully test.
+      let rootTraversal = testModel.traverseFromRoot();
+      assert.isNotEmpty(rootTraversal);
+
+      let searchSpace = new correction.SearchSpace(rootTraversal);
+
+      // VERY artificial distributions.
+      let synthDistribution1 = [
+        {sample: {insert: 't', deleteLeft: 0}, p: 1} // Transform, probability
+      ];
+
+      let synthDistribution2 = [
+        {sample: {insert: 'e', deleteLeft: 0}, p: 0.75}, // Transform, probability
+        {sample: {insert: 'h', deleteLeft: 0}, p: 0.25}
+      ];
+
+      let synthDistribution3 = [
+        {sample: {insert: 'h', deleteLeft: 0}, p: 0.75}, // Transform, probability
+        {sample: {insert: 'n', deleteLeft: 0}, p: 0.25}
+      ];
+
+      searchSpace.addInput(synthDistribution1);
+      searchSpace.addInput(synthDistribution2);
+      searchSpace.addInput(synthDistribution3);
+
+      let iter = searchSpace.getBestMatches();
+      checkResults_teh(iter);
```

Comment:
Current profile:

<img width="473" alt="Screen Shot 2020-08-28 at 2 21 38 PM" src="https://user-images.githubusercontent.com/25213402/91533265-cd993900-e939-11ea-99a9-45c49fbe579b.png">

Note that at present, this test goes (at least) one tier deeper than we'd need to generate a nice set of 3 suggestions.  It's _also_ starting from no input and jumping straight to three characters' worth.

That said, the input distributions are far narrower (and far, far more 'artificial') than we'd see "in practice" when it goes live.

Reply:
> Weird that it's reporting a shorter time.

You are probably seeing CPU cache side effects or the like.

The time for a simple search is a little alarmingly high -- I'd be hoping for a couple of orders of magnitude less, ideally.

Reply:
Yeah.  At the moment, I'm leaving further optimization until later so we can get hands-on testing, but I was hoping to see an order of magnitude less as well.